### PR TITLE
Update of Phase2Digitizer package

### PR DESCRIPTION
**Pull request by suchandradutta**
Porting following PR from 81X 

https://github.com/cms-sw/cmssw/pull/16822

Updated DigitizerUtility::Amplitude object to store the EncodedEventId properly for individual SimHit contributions to the Digi. There was a subtle bug that only the first EncodedEventId was stored. Another bug has also been fixed where the fraction of individual contribution to the digi was calculated wrongly. A few new histograms are added in the validation suite and a few minor bugs corrected

Detailed description of the update reported in Tracker Phase2 Simulation meeting can be found here
https://indico.cern.ch/event/536892/contributions/2385404/attachments/1379749/2096939/DigiStudy_Status_Nov29_2016.pdf

@delaere @atricomi @boudoul @emiglior @kpedro88

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

File: SimTracker/SiPhase2Digitizer/plugins/DigitizerUtility.h [hunk 1]
```diff
@@ -14,61 +14,30 @@
...
 
-    Amplitude( float amp, const PSimHit* hitp, size_t hitIndex, unsigned int tofBin, float frac) :
-    _amp(amp), _frac(1, frac), _hitInfo(new SimHitInfoForLinks(hitp, hitIndex, tofBin)) {
-      // in case of digi from noisypixels
-      // the MC information are removed
-      if (_frac[0] < -0.5) {
-	_frac.pop_back();
-	_hitInfo->trackIds().pop_back();
-      }
+    Amplitude( float amp, float frac) :
```
Comment (davidlange6) [inline SimTracker/SiPhase2Digitizer/plugins/DigitizerUtility.h]: Hi @suchandradutta - you could simply combine all three of these constructors via some default values to the arguments. That would make this cleaner

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 2]
```diff
@@ -944,27 +945,23 @@ void Phase2TrackerDigitizerAlgorithm::digitize(const Phase2TrackerGeomDetUnit* p
...
-	info.hit_counter = sig_data.hitIndex();
-	info.tof_bin     = sig_data.tofBin();
-	info.event_id    = sig_data.eventId();
-	for (unsigned int j = 0; j != sig_data.trackIds().size(); ++j) {
-	  unsigned int tkid = sig_data.trackIds()[j];
-	  float  charge_frac = sig_data.individualampl()[j]/adc;
-	  info.track_map.insert({tkid, charge_frac});
+      if (makeDigiSimLinks_ ) {
+        int j = 0;
+	for (auto l : sig_data.simInfoList()) {
```
Comment (davidlange6) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: what is the typical length of this loop?

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**